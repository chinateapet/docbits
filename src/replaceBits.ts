import { readFile as _readFile, writeFile as _writeFile } from 'fs'
import { relative, dirname, resolve } from 'path'
import { promisify } from 'util'

import { red } from 'chalk'
import globby from 'globby'
import mkdirp from 'mkdirp'

const readFile = promisify(_readFile)
const writeFile = promisify(_writeFile)

type Bits = Map<string | Buffer, string>

interface Options {
	/**
	 * @default "_bits"
	 */
	bitsDirName?: string
	/**
	 * @default "./dist/documents"
	 */
	outputDir?: string
}

/* istanbul ignore next */
/**
 * Replaces markdown bits in Markdown files that live within `docsRoot`.
 * Searches for bits that match the format `${bit-name}` with content
 * from `${options.bitsDirName}/bit-name.md` within the same folder.
 * @param docsRoot defaults to `"./documents"`.
 * @param options
 */
export async function replaceBits(
	docsRoot = './documents',
	{ bitsDirName = '_bits', outputDir = './dist/documents' }: Options = {},
) {
	await mkdirp(outputDir)

	const bitPathsGlob = `${docsRoot}/**/${bitsDirName}/*.md`
	const bits = await loadBits()

	return _replaceBits()

	async function loadBits() {
		const bits: Bits = new Map()
		const bitPaths = await globby(bitPathsGlob)

		await Promise.all(
			bitPaths.map(async (bitPath) => {
				bits.set(
					bitPath.toString(),
					(await readFile(bitPath)).toString(),
				)
			}),
		)

		return bits
	}
	async function _replaceBits() {
		const docPaths = await globby([
			`${docsRoot}/**/*.md`,
			`!${bitPathsGlob}`,
		])

		return Promise.all(docPaths.map(replaceBit))

		async function replaceBit(docPath: string) {
			const bitPattern = /\${([\w\d-]+)}/g
			const docDir = dirname(docPath)
			const contents = (await readFile(docPath)).toString()
			const outputPath = resolve(outputDir, relative(docsRoot, docPath))

			await mkdirp(dirname(outputPath))

			return writeFile(
				outputPath,
				contents.replace(bitPattern, (_m, bitName) => {
					const absoluteDocsRoot = resolve(docsRoot)
					let currentDir = docDir

					while (true) {
						const bitPath = `${currentDir}/${bitsDirName}/${bitName}.md`
						const bit = bits.get(bitPath)
						if (bit) {
							return bit
						}

						const parentDir = resolve(currentDir, '..')
						if (
							currentDir === absoluteDocsRoot ||
							currentDir === parentDir
						) {
							bail()
						}

						currentDir = parentDir
					}

					function bail() {
						throw new Error(
							red(
								`bit \${${bitName}} not found relative to ${relative(
									process.cwd(),
									docDir,
								)}`,
							),
						)
					}
				}),
			)
		}
	}
}
