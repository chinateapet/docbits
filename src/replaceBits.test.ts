import { readFile as _readFile } from 'fs'
import { tmpdir } from 'os'
import { join } from 'path'
import { promisify } from 'util'

import mkdirp from 'mkdirp'

import pkg from '../package.json'

import { replaceBits } from './replaceBits'

const readFile = promisify(_readFile)

describe(replaceBits.name, () => {
	const fixturesPath = 'src/__fixtures__'
	const outputDir = join(tmpdir(), pkg.name)

	mkdirp.sync(outputDir)

	describe('nested bits', () => {
		beforeAll(() =>
			replaceBits(`${fixturesPath}/nested`, {
				outputDir,
			}),
		)

		test('home', async () => {
			const buffer = await readFile(`${outputDir}/home.md`)

			expect(buffer.toString()).toMatchSnapshot()
		})

		test('project-x', async () => {
			const buffer = await readFile(`${outputDir}/project-x/home.md`)

			expect(buffer.toString()).toMatchSnapshot()
		})
	})

	it("throws if bit doesn't exist", () => {
		return expect(
			replaceBits(`${fixturesPath}/missing-bit`, {
				outputDir,
			}),
		).rejects.toThrow('bit ${does-not-exist} not found relative to')
	})
})
