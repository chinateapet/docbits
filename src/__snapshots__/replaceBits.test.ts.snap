// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaceBits nested bits home 1`] = `
"# Home

root table of contents

"
`;

exports[`replaceBits nested bits project-x 1`] = `
"# Project X

table of contents for project-x

"
`;
